fairscale/nn/data_parallel/fully_sharded_data_parallel.py

```diff
@@ -1409,12 +1424,21 @@ def _remove_shard_bwd_hook(fsdp_module: FullyShardedDataParallel) -> None:
                     else:
                         m.assert_state(TrainingState.BACKWARD_PRE)
                 else:
-                    # Unlikely case. When `m` and its children has no params or has params but
-                    # none with `requires_grad==True`, then m's pre-backward and post-backward
-                    # hooks aren't called by autograd. Therefore, it is in IDLE state.
-                    m.assert_state(TrainingState.IDLE)
+                    # When `m` and its children has no params or has params but
+                    # none with `requires_grad==True`, there are two cases:
+                    # 1. output tensors are `requires_grad==True`. In this case,
+                    # pre-backward hook is still registered, so it is in BACKWARD_PRE state.
+                    # 2. output tensors are `requires_grad==False`. In this case,
```

Comment:
Are you sure output tensors have `requires_grad` field? I thought that field is only for params. I am a bit worried that we can see the IDLE state here. That would be a case where the pre-backward callback is not even called?

Reply:
yes, every tensor could have requires_grad field, pre-backward hooks are registered only to output_tensors with requires_grad=True. 

IDLE state is checked here even before this diff. actually it makes sense to me. for some FSDP layers, if there is no params or no any param with requires_grad = true, the output tensors have usually requires_grad=False, so the pre-backward hook is not registered;  for some rare cases, e..g, in test_fsdp_overlap.py, it intentionally set output tensors with requires_grad = true and return, so pre-backward hook is registered

Reply:
I see. Make a lot of sense. Perhaps we can ensure the output of root FSDP's always have requires_grad==True? Will that simplify the case and ensure the pre-back is always called?

Reply:
we do not have to require root FSDP's outputs with requires_grad=True, if all outputs of root FSDP are with requires_grad=False, that is fine, in this case, no backward can be called, so no need to register final backward. 

Here if codes are running at this point, that means root FSDP's pre-backward hook has already been fired and final backward callback was fired.  assert_state(IDLE) is for submodules, not for root module

Reply:
OK, make sense. In that case, should we set the fired flag to True in the beginning of the iteration since we *expect* the final callback not being called? In fact, in that case, none of the backward callbacks will be called. It is like a pure inference case or torch.no_grad() case, right?

Reply:
> OK, make sense. In that case, should we set the fired flag to True in the beginning of the iteration since we _expect_ the final callback not being called?
hmm, which fired flag?

> In fact, in that case, none of the backward callbacks will be called. It is like a pure inference case or torch.no_grad() case, right?
yes

Reply:
sorry, by the "fired flag", I mean the `_post_backward_callback_fired`  boolean. I think I am still confused by this flag. Let's have a chat about it. I will ping you.